libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -571,26 +954,10 @@ MQTTStatus_t MQTT_UpdateStateAck( MQTTContext_t * pMqttContext,
     if( recordIndex < MQTT_STATE_ARRAY_MAX_COUNT )
     {
         newState = MQTT_CalculateStateAck( packetType, opType, qos );
-        shouldDeleteRecord = ( newState == MQTTPublishDone ) ? true : false;
-        isTransitionValid = validateTransitionAck( currentState, newState );
 
-        if( isTransitionValid == true )
-        {
-            updateRecord( records,
-                          recordIndex,
-                          newState,
-                          shouldDeleteRecord );
-
-            status = MQTTSuccess;
-            *pNewState = newState;
-        }
-        else
-        {
-            status = MQTTIllegalState;
-            LogError( ( "Invalid transition from state %s to state %s.",
-                        MQTT_State_strerror( currentState ),
-                        MQTT_State_strerror( newState ) ) );
-        }
+        /* Validate state transition and update state record. */
+        status = updateStateAck( records, recordIndex, packetId, currentState, &newState );
+        *pNewState = newState;
```

Comment:
It looks like now this function will set `pNewState` even if updating the state fails, when previously it would only do so if the return status was MQTTSuccess. However, it won't set it if the return status is MQTTBadParameter (due to the possibility of `pNewState == NULL`). Even though it sets it to MQTTStateNull on error, should we add an `if( status == MQTTSuccess)` check here to only set it on success? That would also mean `updateStateAck` wouldn't need to take a pointer parameter